features/lwipstack/LWIPStack.cpp

```diff
@@ -292,7 +293,20 @@ nsapi_error_t LWIP::socket_open(nsapi_socket_t *handle, nsapi_protocol_t proto)
 nsapi_error_t LWIP::socket_close(nsapi_socket_t handle)
 {
     struct mbed_lwip_socket *s = (struct mbed_lwip_socket *)handle;
-
+#if LWIP_TCP
+    u8_t check_limit = TCP_CLOSE_TIMEOUT;
+    /* Check if TCP FSM is in ESTABLISHED state.
+     * Then give extra time for connection close handshaking  until TIME_WAIT state.
+     * The purpose is to prevent eth/wifi driver stop and  FIN ACK corrupt.
+     * This may happend if network interface disconnect follows immediately after socket_close.*/
+    if (NETCONNTYPE_GROUP(s->conn->type) == NETCONN_TCP && s->conn->pcb.tcp->state == ESTABLISHED) {
+        netconn_shutdown(s->conn, false, true);
+        while (s->conn->pcb.tcp && s->conn->pcb.tcp->state < TIME_WAIT && check_limit) {
+            wait_ms(5);
```

Comment:
`ThisThread::sleep_for(5)` is better - `wait_ms` is on the way to deprecation.

Although, I wonder if you now get an event from lwIP when the state changes, so you can do an `EventFlags::wait(TCP_CLOSE_TIMEOUT)` rather than a loop. Probably no biggie.